src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/FrozenDictionary.AlternateLookup.cs

```diff
@@ -0,0 +1,141 @@
+        /// <returns>The created lookup instance.</returns>
+        /// <exception cref="InvalidOperationException">This instance's comparer is not compatible with <typeparamref name="TAlternateKey"/>.</exception>
+        /// <remarks>
+        /// This instance must be using a comparer that implements <see cref="IAlternateEqualityComparer{TAlternateKey, TKey}"/> with
+        /// <typeparamref name="TAlternateKey"/> and <typeparamref name="TKey"/>. If it doesn't, an exception will be thrown.
+        /// </remarks>
+        public AlternateLookup<TAlternateKey> GetAlternateLookup<TAlternateKey>() where TAlternateKey : notnull, allows ref struct
+        {
+            if (!TryGetAlternateLookup(out AlternateLookup<TAlternateKey> lookup))
+            {
+                ThrowHelper.ThrowIncompatibleComparer();
+            }
+
+            return lookup;
+        }
+
+        /// <summary>
+        /// Gets an instance of a type that may be used to perform operations on a <see cref="FrozenDictionary{TKey, TValue}"/>
+        /// using a <typeparamref name="TAlternateKey"/> as a key instead of a <typeparamref name="TKey"/>.
+        /// </summary>
+        /// <typeparam name="TAlternateKey">The alternate type of a key for performing lookups.</typeparam>
+        /// <param name="lookup">The created lookup instance when the method returns true, or a default instance that should not be used if the method returns false.</param>
+        /// <returns>true if a lookup could be created; otherwise, false.</returns>
+        /// <remarks>
+        /// This instance must be using a comparer that implements <see cref="IAlternateEqualityComparer{TAlternateKey, TKey}"/> with
+        /// <typeparamref name="TAlternateKey"/> and <typeparamref name="TKey"/>. If it doesn't, the method will return false.
+        /// </remarks>
+        public bool TryGetAlternateLookup<TAlternateKey>(out AlternateLookup<TAlternateKey> lookup) where TAlternateKey : notnull, allows ref struct
+        {
+            // The comparer must support the specified TAlternateKey. If it doesn't we can't create a lookup.
+            // Some implementations where TKey is string rely on the length of the input and use it as part of the storage scheme.
+            // That means we can only support TAlternateKeys that have a length we can check, which means we have to special-case
+            // it. Since which implementation we pick is based on a heuristic and can't be predicted by the consumer, we don't
+            // just have this requirement in that one implementation but for all implementations that might be picked for string.
+            // As such, if the key is a string, we only support ReadOnlySpan<char> as the alternate key.
+            if (Comparer is IAlternateEqualityComparer<TAlternateKey, TKey> &&
+                (typeof(TKey) != typeof(string) || typeof(TAlternateKey) == typeof(ReadOnlySpan<char>)))
```

Comment:
Why can't this be generalized to `typeof(TAlternate) == typeof(ReadOnlySpan<TKey>)`?

Reply:
Nevermind, I misunderstood what the check was doing.

Reply:
> Why can't this be generalized to typeof(TAlternate) == typeof(ReadOnlySpan<TKey>)?

That would be e.g. `ReadOnlySpan<string>` not `ReadOnlySpan<char>`.